spec/controller/carts_controller_spec.rb

```diff
@@ -0,0 +1,222 @@
+      end
+    end
+
+    context "user unlogin" do
+      before do
+        get :index
+      end
+
+      it "assigns @carts" do
+        expect(assigns(:carts)).to redirect_to login_url
+      end
+
+      it "redirect login" do
+        expect(response).to redirect_to login_url
+      end
+    end
+  end
+
+  describe "POST #create" do
+    before do
+      session[:user_id] = user.id
+      session[:cart] = {}
+    end
+
+    context "when book exist" do
+      before do
+        @book = book
+        post :create, params: {id: book.id, quantity: 1}
+      end
+
+      it "assigns @book" do
+        expect(assigns(:book)).to eq @book
+      end
+
+      it "add to cart success" do
+        expect(session[:cart][book.id].to_i).to eq 1
+      end
+
+      it "show flash message" do
+        expect(flash[:success]).to eq I18n.t("notice.add_cart_success")
+      end
+
+      it "has a 302 status code" do
+        expect(response).to have_http_status(:found)
+      end
+
+      it "redirect book" do
+        expect(response).to redirect_to book_path(@book)
+      end
+    end
+
+    context "when product does not exist" do
+      before do
+        post :create, params: {id: book.id-1, quantity: 1}
+      end
+
+      it "show flash message" do
+        expect(flash[:danger]).to eq I18n.t("error.not_found_book")
+      end
+
+      it "redirect home" do
+        expect(response).to redirect_to home_path
+      end
+    end
+
+    context "when quantity is over permit range" do
+      before do
+        @book = book
+        post :create, params: {id: book.id, quantity: 4}
+      end
+
+      it "assigns @book" do
+        expect(assigns(:book)).to eq @book
+      end
+
+      it "show flash message" do
+        expect(flash[:danger]).to eq I18n.t("error.quantity.over")
+      end
+    end
+
+    context "when quantity is invalid" do
+      before do
+        session[:user_id] = user.id
+        session[:cart] = {}
+        @book = book
+        post :create, params: {id: book.id, quantity: 0}
+      end
+
+      it "assigns @book" do
+        expect(assigns(:book)).to eq @book
+      end
+
+      it "show flash message" do
+        expect(flash[:danger]).to eq I18n.t("error.quantity.not_valid")
+      end
+
+      it "redirect book" do
+        expect(response).to redirect_to book_path(@book)
+      end
+    end
+  end
+
+  describe "PUT #update" do
+    before do
+      session[:user_id] = user.id
+      session[:cart] = {book.id.to_s => 1}
+    end
+
+    context "when product exist" do
+      before do
+        put :update, params: {id: book.id, quantity: 2}
+      end
+
+      it "update success" do
+        expect(session[:cart][book.id.to_s].to_i).to eq 2
+      end
+
+      it "has a 204 status code" do
+        expect(response).to have_http_status(:no_content)
+      end
+    end
+    context "when product does not exist" do
+      before do
+        put :update, params: {id: book.id + 1, quantity: 2}
```

Comment:
```suggestion
        put :update, params: {id: -1, quantity: 2}
```
ri cho chắc, chứ chắc j +1 là k có mô :v